Document how to use KubeVirt (#528)
coreos/fedora-coreos-docs: modules/ROOT/pages/provisioning-kubevirt.adoc

= Provisioning Fedora CoreOS on KubeVirt

This guide shows how to provision new Fedora CoreOS (FCOS) nodes on any KubeVirt-enabled Kubernetes cluster.

== Prerequisites

Before provisioning an FCOS machine, you must have an Ignition configuration file containing your customizations. If you do not have one, see xref:producing-ign.adoc[Producing an Ignition File].

You also need to have access to a Kubernetes environment with https://kubevirt.io/user-guide/operations/installation/[KubeVirt] installed.

== Referencing the KubeVirt Image

Fedora CoreOS is designed to be updated automatically, with different schedules per stream.

The image for each stream can directly be referenced from the official registry:

----
quay.io/fedora/fedora-coreos-kubevirt

> ----

Reviewer: maybe let's just go ahead and make this a bulleted list and list out the three options:

- `quay.io/fedora/fedora-coreos-kubevirt:stable`
- `quay.io/fedora/fedora-coreos-kubevirt:testing`
- `quay.io/fedora/fedora-coreos-kubevirt:next`

== Launching a VM instance

> Now that you have an image it can be used in your VMI definitions. The following example will use the image from the stable stream.
> 
> You'll also need the Ignition config.

Reviewer: Let's drop the reference to the stream here (we'll select that below) in the "code" and also combine a few sentences:

```suggestion
Given the `quay.io/fedora/fedora-coreos-kubevirt` images you can create a VMI defnition and combine that with an Ignition config to launch a machine.

In the example below...
```

In the example below the Ignition config stored in local file `example.ign` is exposed to the VMI via a Kubernetes Secret.
Learn about various ways to expose userdata to VMIs in the https://kubevirt.io/user-guide/virtual_machines/startup_scripts/#startup-scripts[KubeVirt user guide].

NOTE: If the user prefers, they can use `oc` instead of `kubectl` in the following commands.

.Creating the secret
[source, bash]
----
kubectl create secret generic ignition-payload --from-file=userdata=example.ign
----

.Launching a VM instance referencing the secret
[source, bash]

> ----

Reviewer: Let's let the user choose the stream here in the example:

```
STREAM='stable' # or 'testing' or 'next'
```

> cat << END > vmi.yaml

Reviewer: optional/preference: remove the space

```suggestion
cat <<END > vmi.yaml
```

---
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  name: my-fcos
spec:
  domain:
    devices:
      disks:
      - disk:
          bus: virtio
        name: containerdisk
      - disk:
          bus: virtio
        name: cloudinitdisk
      rng: {}
    resources:
      requests:
        memory: 2048M
  volumes:
  - containerDisk:

>       image: quay.io/coreos/fedora-coreos:stable

Reviewer: ```suggestion
      image: quay.io/coreos/fedora-coreos:${STREAM}
```

= Provisioning Fedora CoreOS on KubeVirt

This guide shows how to provision new Fedora CoreOS (FCOS) nodes on any KubeVirt-enabled Kubernetes cluster.

== Prerequisites

Before provisioning an FCOS machine, you must have an Ignition configuration file containing your customizations. If you do not have one, see xref:producing-ign.adoc[Producing an Ignition File].

You also need to have access to a Kubernetes environment with https://kubevirt.io/user-guide/operations/installation/[KubeVirt] installed.

== Referencing the KubeVirt Image

Fedora CoreOS is designed to be updated automatically, with different schedules per stream.

The image for each stream can directly be referenced from the official registry:

----
quay.io/fedora/fedora-coreos-kubevirt
----

== Launching a VM instance

Now that you have an image it can be used in your VMI definitions. The following example will use the image from the stable stream.

You'll also need the Ignition config.

> Here the Ignition config is exposed to the VMI via a Kubernetes Secret.

Reviewer: ```suggestion
In the example below the Ignition config stored in local file `example.ign` is exposed to the VMI via a Kubernetes Secret.
```

Learn about various ways to expose userdata to VMIs in the https://kubevirt.io/user-guide/virtual_machines/startup_scripts/#startup-scripts[KubeVirt user guide].

> NOTE: You must edit the following commands per your container management system. For example, you must replace `kubectl` with `oc` if you use Openshift.

Reviewer: You can actually use `kubectl` with OpenShift, it just doesn't support everything. Maybe just update the `note` here to say if the user prefers they can use `oc` instead of `kubectl`.

.Creating the secret
[source, bash]
----
kubectl create secret generic ignition-payload --from-file=userdata=example.ign
----

.Launching a VM instance referencing the secret
[source, bash]
----
cat << END > vmi.yaml
---
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  name: my-fcos
spec:
  domain:
    devices:
      disks:
      - disk:
          bus: virtio
        name: containerdisk
      - disk:
          bus: virtio
        name: cloudinitdisk
      rng: {}
    resources:
      requests:
        memory: 2048M
  volumes:
  - containerDisk:
      image: quay.io/coreos/fedora-coreos:stable
    name: containerdisk
  - name: cloudinitdisk
    cloudInitConfigDrive:
      secretRef:
        name: ignition-payload
END
kubectl create -f vmi.yaml
----

Now you should be able to SSH into the instance. If you didn't change the defaults, the
username is `core`.

.Accessing the VM instance using kubectl via ssh
[source, bash]
----
kubectl virt ssh core@my-fcos

> ----

Reviewer: I would just drop this since it requires a plugin being installed for kubectl.

.Accessing the VM instance using virtctl via ssh
[source, bash]
----
virtctl ssh core@my-fcos

> ----

Reviewer: would be nice if we could link to some docs about `virtctl` here, but I don't there really are any. Could you find some?

Author: I can only see https://kubevirt.io/user-guide/operations/virtctl_client_tool/ which is decent and will help to install virtctl for users who don't have it installed. You've probably already seen this option but there really isn't any other good option.

Reviewer: That looks great. Let's use that.

== Mirroring the image for use in private registries

> If a private registry in air-gapped installations is used, the image can be mirrored to that registry.

Reviewer: ```suggestion
If a private registry in air-gapped installations is used, the image can be mirrored to that registry using https://github.com/containers/skopeo[`skopeo`].
```

> .Mirroring a fcos image from stable stream

Reviewer: ```suggestion
.Mirroring a stable stream FCOS image
```

[source, bash]
----

> skopeo copy docker://quay.io/coreos/fedora-coreos:stable docker://myregistry.io/myorg/fedora-coreos:stable

Reviewer: ```suggestion
skopeo copy docker://quay.io/coreos/fedora-coreos-kubevirt:stable docker://myregistry.io/myorg/fedora-coreos-kubevirt:stable
```
